src/GraphQL/Types/Scalars/IntGraphType.cs

```diff
@@ -12,25 +12,56 @@ public class IntGraphType : ScalarGraphType
         /// <inheritdoc/>
         public override object ParseLiteral(IValue value) => value switch
         {
-            IntValue intValue => intValue.Value,
-            LongValue longValue => int.MinValue <= longValue.Value && longValue.Value <= int.MaxValue ? (int?)longValue.Value : null,
-            BigIntValue bigIntValue => int.MinValue <= bigIntValue.Value && bigIntValue.Value <= int.MaxValue ? (int?)bigIntValue.Value : null,
-            _ => null
+            IntValue intValue => BoxValue(intValue.Value),
+            LongValue longValue => checked((int)longValue.Value),
+            BigIntValue bigIntValue => checked((int)bigIntValue.Value),
+            NullValue _ => null,
+            _ => ThrowLiteralConversionError(value)
         };
 
-        /// <inheritdoc/>
-        public override object ParseValue(object value) => ValueConverter.ConvertTo(value, typeof(int));
-
         /// <inheritdoc/>
         public override bool CanParseLiteral(IValue value) => value switch
         {
             IntValue _ => true,
             LongValue longValue => int.MinValue <= longValue.Value && longValue.Value <= int.MaxValue,
             BigIntValue bigIntValue => int.MinValue <= bigIntValue.Value && bigIntValue.Value <= int.MaxValue,
+            NullValue _ => true,
             _ => false
         };
 
         /// <inheritdoc/>
-        public override IValue ToAST(object value) => new IntValue(Convert.ToInt32(value));
+        public override object ParseValue(object value) => value switch
+        {
+            int _ => value,
+            null => null,
+            sbyte sb => checked((int)sb),
+            byte b => checked((int)b),
+            short s => checked((int)s),
+            ushort us => checked((int)us),
+            uint ui => checked((int)ui),
+            long l => checked((int)l),
+            ulong ul => checked((int)ul),
+            BigInteger bi => (int)bi,
+            _ => ThrowValueConversionError(value)
+        };
+
+        private static readonly object _boxedNeg1 = -1;
+        private static readonly object _boxed0 = 0;
+        private static readonly object _boxed1 = 1;
+        private static readonly object _boxed2 = 2;
+        private static readonly object _boxed3 = 3;
+        private static readonly object _boxed4 = 4;
+        private static readonly object _boxed5 = 5;
+        private static object BoxValue(int value) => value switch
```

Comment:
From time to time I thought about doing this, but gave up this thought. This looks a little odd and adds processing time. We need to look at the problem more broadly - #2237 
![изображение](https://user-images.githubusercontent.com/21261007/109619196-88a0ca00-7b49-11eb-938c-7dd0728b6bd3.png)

Reply:
I've seen this before; I think it's generally done the way I did it -- a very small (5-10) number of literals are pre-boxed.  I could be wrong.

I only added this code for `IntGraphType` as it should be the most common numeric type.  However, it's only useful for input literals so maybe it's not that valuable.  Keep or remove?  I did add tests for this, by the way.

Reply:
> I've seen this before

Maybe for index in response path? There it is more useful because the index increases monotonically from 0. Therefore, the "boxed cache" hit rate is high. In the case of scalars, we work with arbitrary values and we have no idea what values are used more often. I would remove this code.

Reply:
Done